src/components/action-popover/action-popover-item/action-popover-item.component.tsx

```diff
@@ -139,19 +143,106 @@ export const ActionPopoverItem = ({
   const mouseEnterTimer = useRef<NodeJS.Timeout | null>(null);
   const mouseLeaveTimer = useRef<NodeJS.Timeout | null>(null);
 
+  const currentSubmenuAlignment = () => {
+    if (submenuPosition === "left") {
+      if (leftAlignmentState) {
+        return "right";
+      }
+      return "left";
+    }
+    return "right";
+  };
+
+  function getContainerPosition(
+    itemRef: React.RefObject<HTMLElement>,
+    containerSubmenuRef: React.RefObject<HTMLElement>,
+    containerPlacement: "bottom" | "top"
+  ): ContainerPosition | undefined {
+    /* istanbul ignore if */
+    if (!itemRef.current || !containerSubmenuRef.current) return undefined;
+
+    const { offsetWidth: submenuWidth } = containerSubmenuRef.current;
+
+    const horizontalAlignmentValue = currentSubmenuAlignment() === "left";
+    const leftValue = horizontalAlignmentValue ? -submenuWidth : "auto";
+    const rightValue = horizontalAlignmentValue ? "auto" : -submenuWidth;
+    const yPositionName = containerPlacement === "top" ? "bottom" : "top";
+
+    return {
+      left: leftValue,
+      [yPositionName]: "calc(-1 * var(--spacing100))",
+      right: rightValue,
+    };
+  }
+
+  const retrieveVariantType = () => {
+    if (!iconState && !caretState && !icon && !submenu) {
+      return 1;
+    }
+    if (!iconState && caretState && !icon && !submenu) {
+      return 2;
+    }
+    if (!iconState && caretState && !icon && submenu) {
+      return 3;
+    }
+    if (iconState && !caretState && !icon && !submenu) {
+      return 4;
+    }
+    if (iconState && !caretState && icon && !submenu) {
+      return 5;
+    }
+    if (iconState && caretState && icon && submenu) {
+      return 6;
+    }
+    if (iconState && caretState && icon && !submenu) {
+      return 7;
+    }
+    if (iconState && caretState && !icon && submenu) {
+      return 8;
+    }
+    if (iconState && caretState && !icon && !submenu) {
+      return 9;
+    }
+    return 1;
+  };
+
   useEffect(() => {
     if (!isOpenPopover) {
       setOpen(false);
     }
   }, [isOpenPopover]);
 
+  useEffect(() => {
```

Comment:
**suggestion**: as far as I can tell this `useEffect` is the only place where `setIconState` or `setCaretState` are actually used. If that's the case I don't understand why we need all the complication of passing down `iconState`, `caretState` and the 2 setters down as props just to achieve this - you can just remove all of them and set use `icon` everywhere you're using `iconState` in this child component, and `submenu` everywhere you're using `caretState`. Am I missing something which means we can't do this?

Actually there is one difference between the two - your version will `setIconState` to `true` if `icon` ever becomes `true`, but won't then set it back to `false` if `icon` goes back to being false. (Same with `submenu` and `setCaretState`.) I doubt it matters much in practice as these props aren't likely to change dynamically over the popover menu's lifecycle - but they could, in which case I think your version is bugged. Anyway, if I'm right about that it can be easily fixed, but it would be much better to do the simplification I've suggested.

Please let me know if I'm overlooking something which means we have to do it the complicated way!

Reply:
Hi Robin, so `setIconState` and `setCaretState` which set `iconState` and `caretState` to true respectively is how we've made design changes based on the value of siblings. So let's say we have x8 children and the first has an `icon`, `setIconState` will be triggered and will be changed to `true`, and will not go back to `false`. This is the desired original behaviour and not a bug imo, as now we can make design changes knowing a sibling has an `Icon`. 

<img width="215" alt="Screenshot 2023-06-21 at 13 46 13" src="https://github.com/Sage/carbon/assets/114918852/d0731add-c02e-4788-9f69-2414ac01e8b6">

The above screenshot is a good example of this, the first child has an `icon` so `iconState` is true, and remains true even though the next three children do not have an `icon`. This means those three children know a sibling has an icon, so padding is applied to each text to ensure padding is the same, which matches designs. This was my solution to make changes based on the value of siblings 👍. Happy to explore alternatives if you're not happy with this, though, or if you still feel it's incorrect.

Reply:
Thanks, I'll take another look at your code but I think this is raising something important I'd missed - that these are being passed down to all children so any one of them can affect all the others. That definitely makes it more complex than I was initially thinking. I'll have to have a think about if there's any other way which removes the need to have all these `useEffect`s (there may not be though!)